// import-sort-config/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assign = require("core-js/library/fn/object/assign");
const cosmiconfig = require("cosmiconfig");
const minimatch = require("minimatch");
const resolve = require("resolve-from").silent;
exports.DEFAULT_CONFIGS = {
    ".js, .jsx, .es6, .es": {
        parser: "babylon",
        style: "eslint",
    },
    ".ts, .tsx": {
        parser: "typescript",
        style: "eslint",
    },
};
function getConfig(extension, directory, defaultConfigs = exports.DEFAULT_CONFIGS) {
    const defaultConfig = getConfigForExtension(defaultConfigs, extension);
    let packageConfig;
    if (directory) {
        packageConfig = getConfigFromDirectory(directory, extension);
    }
    const actualConfig = mergeConfigs([defaultConfig, packageConfig]);
    if (!actualConfig) {
        return;
    }
    const resolvedConfig = resolveConfig(actualConfig, directory);
    return resolvedConfig;
}
exports.getConfig = getConfig;
function getConfigFromDirectory(directory, extension) {
    const packageConfigs = getAllConfigsFromDirectory(directory);
    if (!packageConfigs) {
        return;
    }
    return getConfigForExtension(packageConfigs, extension);
}
function getConfigForExtension(configs, extension) {
    const foundConfigs = Object.keys(configs).map(joinedGlobs => {
        const globs = joinedGlobs.split(",").map(rawGlob => rawGlob.trim());
        const config = configs[joinedGlobs];
        if (globs.some(glob => minimatch(extension, glob))) {
            return config;
        }
    });
    return mergeConfigs(foundConfigs);
}
function getAllConfigsFromDirectory(directory) {
    const configsLoader = cosmiconfig("importsort", {
        sync: true,
        packageProp: "importSort",
        rcExtensions: true,
    });
    try {
        const configsResult = configsLoader.load(directory);
        if (!configsResult) {
            return;
        }
        return configsResult.config;
    }
    catch (e) {
        return;
    }
}
function mergeConfigs(rawConfigs) {
    const configs = rawConfigs.filter(rawConfig => !!rawConfig);
    if (configs.length === 0) {
        return;
    }
    return configs.reduce((previousConfig, currentConfig) => {
        if (!currentConfig) {
            return previousConfig;
        }
        const config = assign({}, previousConfig);
        if (currentConfig.parser) {
            config.parser = currentConfig.parser;
        }
        if (currentConfig.style) {
            config.style = currentConfig.style;
        }
        if (currentConfig.options) {
            config.options = currentConfig.options;
        }
        return config;
    });
}
function resolveConfig(config, directory) {
    const resolvedConfig = {
        config,
    };
    if (config.parser) {
        resolvedConfig.parser = resolveParser(config.parser, directory);
    }
    if (config.style) {
        resolvedConfig.style = resolveStyle(config.style, directory);
    }
    return resolvedConfig;
}
function resolveParser(module, directory) {
    if (module.indexOf(".") === 0) {
        return resolveModule(module, directory);
    }
    if (module.indexOf("import-sort-parser") === 0) {
        return resolveModule(module, directory);
    }
    return resolveModule(`import-sort-parser-${module}`, directory);
}
function resolveStyle(module, directory) {
    if (module.indexOf(".") === 0) {
        return resolveModule(module, directory);
    }
    if (module.indexOf("import-sort-style-") === 0) {
        return resolveModule(module, directory);
    }
    return resolveModule(`import-sort-style-${module}`, directory);
}
function resolveModule(module, directory) {
    if (directory) {
        const path = resolve(directory, module);
        if (path) {
            return path;
        }
    }
    const path = resolve(__dirname, module);
    if (path) {
        return path;
    }
}
